### PR TITLE
Z Moves

### PR DESCRIPTION
**Pull request by ExpoSeed**
Z MOVES IS MINE NOW

### CODE REVIEW

File: data/battle_anim_scripts.s [hunk 1]
```diff
@@ -24406,7 +24481,7 @@ RainbowEndureEffect:
 	delay 0x3
 	launchtemplate gYellowEndureEnergySpriteTemplate 0x2 0x4 0x0 0x1c 0x1a 0x3
 	delay 0x3
-	return
+	return
```
Comment (LOuroboros) [inline data/battle_anim_scripts.s:24484]:
```suggestion
	return
```

File: data/battle_anim_scripts.s [hunk 2]
```diff
@@ -24500,3 +24575,5632 @@ Special_CriticalCaptureBallThrow:
...
+	playsewithpan SE_M_EXPLOSION, SOUND_PAN_ATTACKER
+	launchtemplate gSlideMonToOffsetSpriteTemplate 0x2 0x5 ANIM_TARGET 0xff00 0x15 0x0 0x4
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x0 0xc 0xffe8 0xfff0 0x18 	@ 12, -48, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x0 0x10 0xffda 0xfff6 0x18 	@ 16, -16, -10, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x1 0xe 0xffec 0xffee 0x18 	@ 14, -52, -18, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x1 0xc 0xffdc 0xfff0 0x18 	@ 12, -32, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x0 0xc 0xffd0 0xfff0 0x18 		@ 12, -24, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x0 0x10 0xfff0 0xfff6 0x18 		@ 16, -38, -10, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x1 0xe 0xffcc 0xffee 0x18 		@ 14, -20, -18, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x82 0x6 ANIM_TARGET 0x1 0xc 0xffe0 0xfff0 0x18 		@ 12, -36, -16, 24
```
Comment (LOuroboros) [inline data/battle_anim_scripts.s:24813]: The comments should all be aligned when possible, imo.

File: data/battle_anim_scripts.s [hunk 3]
```diff
@@ -24500,3 +24575,5632 @@ Special_CriticalCaptureBallThrow:
...
+	playsewithpan SE_M_EXPLOSION, SOUND_PAN_TARGET
+	launchtemplate gSlideMonToOffsetSpriteTemplate 0x2 0x5 ANIM_TARGET 0xff00 0x15 0x0 0x4
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x0 0xc 0xffe8 0xfff0 0x18 	@ 12, -48, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x0 0x10 0xffda 0xfff6 0x18 	@ 16, -16, -10, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x1 0xe 0xffec 0xffee 0x18 	@ 14, -52, -18, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x1 0xc 0xffdc 0xfff0 0x18 	@ 12, -32, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x0 0xc 0xffd0 0xfff0 0x18 		@ 12, -24, -16, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x0 0x10 0xfff0 0xfff6 0x18 		@ 16, -38, -10, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x1 0xe 0xffcc 0xffee 0x18 		@ 14, -20, -18, 24
+	launchtemplate gDirtPlumeSpriteTemplate 0x2 0x6 ANIM_TARGET 0x1 0xc 0xffe0 0xfff0 0x18 		@ 12, -36, -16, 24
```
Comment (LOuroboros) [inline data/battle_anim_scripts.s:25120]: The comments should all be aligned when possible, imo. There's no reason for these 4 not to be aligned with the 4 above them.

File: data/battle_anim_scripts.s [hunk 4]
```diff
@@ -24500,3 +24575,5632 @@ Special_CriticalCaptureBallThrow:
...
+	delay 0x2
+	launchtemplate gContinentalCrushNeedleArmSpriteTemplate 0x2 0x5 ANIM_ATTACKER 0x0 0x0 0xffce 0x20 	@ 0, -50
+	delay 0x2
+	launchtemplate gContinentalCrushNeedleArmSpriteTemplate 0x2 0x5 ANIM_ATTACKER 0x0 0x22 0xffde 0x20 	@ +34, -34
+	delay 0x2
+	launchtemplate gContinentalCrushNeedleArmSpriteTemplate 0x2 0x5 ANIM_ATTACKER 0x0 0x2e 0x0 0x20 		@ +46, +0
+	delay 0x2
+	launchtemplate gContinentalCrushNeedleArmSpriteTemplate 0x2 0x5 ANIM_ATTACKER 0x0 0x20 0x20 0x20 	@ +32, +32
+	delay 0x2
+	return
```
Comment (LOuroboros) [inline data/battle_anim_scripts.s:25171]: The comments should all be aligned when possible, imo. There's no reason for the comments that are misaligned, such as Line 25155's or Line 25167 to be misaligned compared to the majority of the others.

File: data/battle_scripts_1.s [hunk 5]
```diff
@@ -7927,6 +7930,122 @@ BattleScript_JabocaRowapBerryActivate_Dmg:
...
+	hitanimation BS_TARGET
+	waitstate
+	healthbarupdate BS_TARGET
+	datahpupdate BS_TARGET
+	critmessage
+	waitmessage 0x40
+	resultmessage
+	waitmessage 0x40
+	setterrain BattleScript_TryFaint
+@@@ TODO - 'restore' bg @@@
```
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:8043]: Unfortunately I don't know how to do this as I am very bad with battle scripts

File: data/battle_scripts_1.s [hunk 6]
```diff
@@ -369,6 +369,9 @@ gBattleScriptsForMoveEffects::
 	.4byte BattleScript_EffectSleepHit                @ EFFECT_SLEEP_HIT
 	.4byte BattleScript_EffectAttackerDefenseDownHit  @ EFFECT_ATTACKER_DEFENSE_DOWN_HIT
 	.4byte BattleScript_EffectHit                     @ EFFECT_BODY_PRESS
+	.4byte BattleScript_EffectExtremeEvoboost         @ EFFECT_EXTREME_EVOBOOST
+	.4byte BattleScript_EffectTerrainHit              @ EFFECT_DAMAGE_SET_TERRAIN
+	.4byte BattleScript_EffectHit
```
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:374]: There isn't, that entry doesn't correspond with any effect

File: data/battle_scripts_1.s (hunk 5 above)
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:8043]: I'll leave that as a job for a future PR

File: data/battle_scripts_1.s (hunk 6 above)
[the remark this answers is not in the document]
Comment (LOuroboros) [reply data/battle_scripts_1.s:374]: But the last entry of the list *is* supposed to correspond with a move effect, no? 🤔 
https://github.com/rh-hideout/pokeemerald-expansion/blob/battle_engine/data/battle_scripts_1.s#L405

[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:374]: not in the code i have https://github.com/ExpoSeed/pokeemerald/blob/BE/z-moves/include/constants/battle_move_effects.h#L358

[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:374]: perhaps this will be addressed once i merge

[the remark this answers is not in the document]
Comment (LOuroboros) [reply data/battle_scripts_1.s:374]: > not in the code i have [ExpoSeed/pokeemerald@`BE`/z-moves/include/constants/battle_move_effects.h#L358](https://github.com/ExpoSeed/pokeemerald/blob/BE/z-moves/include/constants/battle_move_effects.h?rgh-link-date=2021-12-16T22%3A13%3A45Z#L358)

That constant is effectively in the BE outside of this PR as well, yet, you don't see an entry for it in gBattleScriptsForMoveEffects :eyes: ...

File: src/battle_script_commands.c [hunk 7]
```diff
@@ -7053,12 +7074,12 @@ static bool32 HasAttackerFaintedTarget(void)
...
 }
 
-static void HandleTerrainMove(u32 moveEffect)
+static void HandleTerrainMove(u16 move)
 {
     u32 statusFlag = 0;
     u8 *timer = NULL;
-
-    switch (moveEffect)
+
```
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply src/battle_script_commands.c:7081]: I assume you didn't mean to remove the `switch` statement

**Comment by AsparagusEduardo** [comment]
We need a new UI icon lol

File: data/battle_scripts_1.s [hunk 8]
```diff
@@ -7927,6 +7930,122 @@ BattleScript_JabocaRowapBerryActivate_Dmg:
...
+	printstring STRINGID_ZMOVEUNLEASHED
+	waitmessage 0x40
+	return
+
+BattleScript_ZMoveActivateStatus::
+	printstring STRINGID_ZPOWERSURROUNDS
+	playanimation BS_ATTACKER, B_ANIM_ZMOVE_ACTIVATE, NULL
+	setzeffect
+@	printstring STRINGID_ZMOVEUNLEASHED
+@	waitmessage 0x40
```
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply data/battle_scripts_1.s:9221]: I've removed the comment

File: src/battle_z_move.c [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+            return FALSE;
+        
+        if (!IS_MOVE_STATUS(chosenMove) && AI_CalcDamage(chosenMove, battlerAtk, battlerDef, FALSE) >= gBattleMons[battlerDef].hp)
+            return FALSE;   // don't waste damaging z move if can otherwise faint target
+        
+        return TRUE;
+    }
+    
+    return FALSE;
+}
```
[the remark this answers is not in the document]
Comment (ExpoSeed) [reply src/battle_z_move.c:730]: I'm putting it in battle_ai_util instead, which has all the other ShouldUseFunctions
```c
bool32 ShouldTryOHKO(u8 battlerAtk, u8 battlerDef, u16 atkAbility, u16 defAbility, u32 accuracy, u16 move);
bool32 ShouldUseRecoilMove(u8 battlerAtk, u8 battlerDef, u32 recoilDmg, u8 moveIndex);
u16 GetBattlerSideSpeedAverage(u8 battler);
bool32 ShouldAbsorb(u8 battlerAtk, u8 battlerDef, u16 move, s32 damage);
bool32 ShouldRecover(u8 battlerAtk, u8 battlerDef, u16 move, u8 healPercent);
bool32 ShouldSetScreen(u8 battlerAtk, u8 battlerDef, u16 moveEffect);
bool32 ShouldPivot(u8 battlerAtk, u8 battlerDef, u16 defAbility, u16 move, u8 moveIndex);
```

**Comment by ExpoSeed** [comment]
Ready for review. I took notice of the fact that Z-Growl appears to be kinda wonky and print the wrong string but I do not have the knowledge to fix it

**Pull request closed without merging**